app/models/talentlms/facade.rb

```diff
@@ -10,42 +10,111 @@ def initialize
+      raise e unless e.message.include?('The requested user does not exist')
+    end
+
+    def talentlms_user_record_by_username(username)
+      result = @api.post('users', { username: username })
+      result
+    rescue RuntimeError => e
+      raise e unless e.message.include?('The requested user does not exist')
+    end
+
+    def lms_username(user)
+      "#{ENV['RAILS_ENV']}_#{user.id + Integer(ENV.fetch('DEV_OFFSET', 0))}"
+    end
+
+    def lms_email(user)
+      email_address = "#{lms_username(user)}@#{ENV['FQDN']}"
+      email_address = user.talent_lms_email if user.talent_lms_email.present?
+      email_address
+    end
+
     # Login user to TalentLMS
     #
     # @@param user [User] the user
     # @return [String] URL to redirect user to
     def login(user)
       login = Login.find_by(user: user)
-      if login.nil?
-        result = create_account(user)
-      else
-        result = @api.post('userlogin', { login: login.login, password: login.password })
-      end
+
+      # We may have a local talent record for this user, but we need to run it
+      # through the sync process to ensure that the email address that we have
+      # on record matches the email address that TalentLMS has for this user.
+      result = sync_lms_account(user, login)
       result['login_key']
     end
 
+    def sync_lms_account(user, login)
+      username = lms_username(user)
+      email_address = lms_email(user)
+
+      # If we have a local login record, we have a talent account id. Check here first for the talent record
+      result ||= talentlms_user_record_by_id(login.lms_user_id) if login.present?
+      # The local login record does not exist OR Talent does not have an account for this id.
+      # Check for a Talent account with the assocaited email address
+      result ||= talentlms_user_record_by_email(email_address)
+      # Talent does not have a record associated with this email address, check the default username. If we generated an
+      # account for this user prior to allowing emails to be set, we should be able to find it this way.
+      result ||= talentlms_user_record_by_username(username)
+      # Talent does not have a record associated with this user, create an account for them.
+      result ||= create_account(user)
+
+      # Update the talent_lms_email on the local user if it differs from the email address in TalentLMS
+      if result.present? && result['email'] != user.talent_lms_email
+        user.talent_lms_email = result['email']
+        user.save!
+      end
+
+      # If we want to update the email in Talent to match the local account instead of updating the local email
+      # address to match what we are seeing in Talent. The line below will update the account in TalentLMS.
+      # Leaving this here in case we want to use this behavior.
+      ## result = @api.post('edituser', { user_id: result['id'], email: email_address }) if result.present? && result['email'] != email_address
```

Comment:
This is a good question.  So if we have an email and Talent has an email, but they don't match, we preference our local email right now?  Would we ever be able to identify the user in talent if they had an account in Talent that wasn't created by us and our local email didn't match?

Reply:
If we have an email and Talent has an email and they don't match, but the accounts were matched (by matching the `Talentlms::Login.lms_user_id` record to the TalentLMS account id or the `Talentlms::Login.login` record to the TalentLMS account username. The current setup will update the `User.talent_lms_email` record to what we are getting from the TalentLMS account email. We could push an edit to the TalentLMS account so it uses our stored email, but I defaulted to this way to keep any changes within our environment.

Regarding identifying a user in talent if we didn't create it and our email didn't match, that can be done but would require being able to set/edit what we have stored for the username (`Talentlms::Login.login`) or the TalemtLMS account id (`Talentlms::Login.lms_user_id`)